### ISSUE TITLE
Cost function in VAE example

### ISSUE THREAD

**Issue by atabakd** [fchollet/keras#3169]
The kl part of the cost function in the vae example seems to be different than the one from the authors. In particular, it should be : 

`- 0.5 * K.mean(1 + 2 * z_log_std - K.square(z_mean) -K.square(K.exp(z_log_std)), axis=-1)`

Or we can simply imagine the network is estimating z_log_std_squared and change the sampling function to:

`return z_mean + K.sqrt(K.exp(z_log_std)) * epsilon`

**Comment by atabakd** [comment]
Moreover, changing the reconstruction part of the code to the line below has immensely  (changed the scope of loss values / improved performance)

` xent_loss = \
            -tf.reduce_sum(x * tf.log(1e-10 + x_decoded_mean)
                           + (1-x) * tf.log(1e-10 + 1 - x_decoded_mean),1)`

There must be something going on with the `objectives.binary_crossentropy`

**Comment by michielree** [comment]
I came here to report the same thing :)

Another thing seems to be missing from the KL loss: the effective std of Z is z_log_std * epsilon_std, and that's omitted atm. I believe the correct form would be:

` -0.5 * K.mean(1 + 2 * (z_log_std + np.log(epsilon_std)) - K.square(z_mean) + (epsilon_std ** 2) * K.exp(2 * z_log_std)`

I'd be more than willing to set up a pull request, would be my first OSS contribution so someone telling me how to proceed would be great.

Cheers!

**Comment by michielree** [comment]
Fixed in https://github.com/fchollet/keras/pull/3220
